src/lsm/block_free_set.zig

```diff
@@ -1,37 +1,642 @@
+        }
+    }
+
+    var buf = try std.testing.allocator.alloc(u8, set.encodeSize());
+    defer std.testing.allocator.free(buf);
+
+    try std.testing.expectEqual(buf.len % 8, 0);
+    set.encode(buf);
+    var set2 = try BlockFreeSet.decode(buf, std.testing.allocator);
+    defer set2.deinit(std.testing.allocator);
+
+    try expectBlockFreeSetEqual(set, set2);
+}
+
+fn expectBlockFreeSetEqual(a: BlockFreeSet, b: BlockFreeSet) !void {
+    try expectBitsetEqual(a.blocks, b.blocks);
+    try expectBitsetEqual(a.index, b.index);
+}
+
+fn expectBitsetEqual(a: DynamicBitSetUnmanaged, b: DynamicBitSetUnmanaged) !void {
+    try std.testing.expectEqual(a.bit_length, b.bit_length);
+    var i: usize = 0;
+    while (i < bitsetMasksLen(a)) : (i += 1) try std.testing.expectEqual(a.masks[i], b.masks[i]);
+}
+
+test "BitsetEncoder" {
+    const literals = 2;
+    const runs = 2;
+    const run_lengths = [_]usize{5, 3};
+    const total_blocks = 64 * (literals + run_lengths[0] + run_lengths[1]);
+    const encoded = mem.sliceAsBytes(&[_]u64{
+        total_blocks |
+        (runs + literals) << 32,
+        // run, literal, run, literal
+        0b00000000_00000000_00000000_00000000_00000000_00000000_00000000_00000101,
+        // run of ones, run of zeroes
+        0b00000000_00000000_00000000_00000000_00000000_00000000_00000000_00000001,
+        // run of 5 words, run of 3 words
+        (run_lengths[0] - 1) |
+        (run_lengths[1] - 1) << 8,
+        // literals:
+        0b10101010_10101010_10101010_10101010_10101010_10101010_10101010_10101010,
+        0b01010101_01010101_01010101_01010101_01010101_01010101_01010101_01010101,
+    });
+
+    const decoded = [_]u64{
+        0b11111111_11111111_11111111_11111111_11111111_11111111_11111111_11111111, // run 1
+        0b11111111_11111111_11111111_11111111_11111111_11111111_11111111_11111111,
+        0b11111111_11111111_11111111_11111111_11111111_11111111_11111111_11111111,
+        0b11111111_11111111_11111111_11111111_11111111_11111111_11111111_11111111,
+        0b11111111_11111111_11111111_11111111_11111111_11111111_11111111_11111111,
+        0b10101010_10101010_10101010_10101010_10101010_10101010_10101010_10101010, // literal 1
+        0b00000000_00000000_00000000_00000000_00000000_00000000_00000000_00000000, // run 2
+        0b00000000_00000000_00000000_00000000_00000000_00000000_00000000_00000000,
+        0b00000000_00000000_00000000_00000000_00000000_00000000_00000000_00000000,
+        0b01010101_01010101_01010101_01010101_01010101_01010101_01010101_01010101, // literal 2
+    };
+
+    var got_decoded = try BitsetEncoder(.immutable).decode(encoded, std.testing.allocator);
+    defer got_decoded.deinit(std.testing.allocator);
+
+    try std.testing.expectEqual(total_blocks, got_decoded.bit_length);
+    try std.testing.expectEqual(decoded.len, bitsetMasksLen(got_decoded));
+    try std.testing.expectEqualSlices(u64, &decoded, got_decoded.masks[0..decoded.len]);
```

Comment:
It would be so good to do just one more final roundtrip to encode `decoded_actual` and check that we arrive back at `encoded_expect`! 😄